### PR TITLE
routing

### PR DESCRIPTION
**Pull request by MarkHaoxiang**
1. Bug fixes wrt. routing
2. Doesn't crash anymore when trucks reach final destination
3. Current truck container
4. https://miro.medium.com/max/875/1*rX7jt8J61Pyeu6Qb98mXUg.jpeg Behavior discussion: consider truck with route [5] at node 3. Do we path upwards (most optimal) or go directly (current implementation, allows full path specification). I would recommend adding an option in JSON to force the first behavior, and default the second but that's also more work.

### CODE REVIEW

File: backend/simulation/draw/visualiser.py [hunk 1]
```diff
@@ -1,3 +1,5 @@
+# mypy: ignore-errors
```
Comment (sam1penny) [inline backend/simulation/draw/visualiser.py]: ?
seems to work fine without this (just tested)

File: backend/test/test_env.py [hunk 2]
```diff
@@ -9,5 +9,12 @@ def test_setup(self) -> None:
         env = Env()
         env.reset("../shared/example.json")
 
+    def test_rollout_basic(self) -> None:
+        env = Env()
+        env.reset("test/test_json/test_1.json")
+
+        for _ in range(128):
+            env.step({10001: 10.0})
```
Comment (sam1penny) [inline backend/test/test_env.py:17]: What is this test testing? just that it doesn't crash?

File: backend/simulation/realm/graph.py [hunk 3]
```diff
@@ -71,12 +77,18 @@ def addOutgoing(self, e: Edge) -> None:
...
             raise NotImplementedError
 
 class Junction(Node):
 
     def entry(self, truck: Truck) -> None:
+        super().entry(truck)
+        if truck.done():
+            return
+        self._trucks.pop()
         self._outgoing[self._routing_table[truck.destination]].entry(truck)
```
Comment (sam1penny) [inline backend/simulation/realm/graph.py]: is it a good idea to:
(a) let trucks finish in a junction
(b) store these trucks forever in the junction if they do finish there
?

**Review by sam1penny — COMMENTED** [review]
Left a couple of small comments. Didn't look in depth at the floyd-marshall algorithm but I assume it's correct :)

Did we also agree on changing the way trucks store routes (to not pop/mutate the route, but store a current index). Or if we save this for later, at least use a Deque and store routes in the correct order (otw code is a lot less readable).

Let me know if you disagree with anything.

**Comment by sam1penny** [comment]
In regard to (4), most optimal would be better IMO, however it might be worth putting that task in the backlog, not essential for the MVP.

File: backend/simulation/draw/visualiser.py (hunk 1 above)
Comment (MarkHaoxiang) [reply backend/simulation/draw/visualiser.py]: Strange, broke locally. Might be something wrong with my installation of pygame (error surface.Surface doesn't match Surface).

File: backend/test/test_env.py (hunk 2 above)
Comment (MarkHaoxiang) [reply backend/test/test_env.py:17]: Yes. TODO: Should add more checks, ran out of time.

File: backend/simulation/realm/graph.py (hunk 3 above)
Comment (MarkHaoxiang) [reply backend/simulation/realm/graph.py]: a) I think this should be checked in the configuration file, don't provide a configuration which terminates at a junction.
b) Probably best to delete trucks and just add a "done" id mapping in observation dictionary to save on memory usage. I don't think this is necessary for MVP.

Comment (MarkHaoxiang) [reply backend/simulation/realm/graph.py]: In reply to 4) delete lines 50-53 for most optimal behavior. I don't mind either way.

Comment (sam1penny) [reply backend/simulation/realm/graph.py]: Cool - my point was that we are currently allowing (a) and (b) to occur ( the highlighted code is within the `Junction` class)

File: backend/simulation/realm/truck.py [hunk 4]
```diff
@@ -47,9 +46,9 @@ def current_truck_container(self) -> TruckContainer:
 
     @property
     def destination(self) -> int:
-        if len(self._route) == 0:
-            return -1
-        return self._route[-1]
+        if self.done():
+            return False
+        return self._route[self._route_index]
```
Comment (sam1penny) [inline backend/simulation/realm/truck.py]: This shouldn't type check...

**Pull request merged**